src/Castle.Core/DynamicProxy/Generators/MethodSignatureComparer.cs

```diff
@@ -129,19 +129,24 @@ private bool EqualSignatureTypes(Type x, Type y, MethodInfo xm = null, MethodInf
 			{
 				if (!x.Equals(y))
 				{
-					// This enables covariant method returns for .NET 5 and newer.
-					// No need to check for runtime support, since such methods are marked with a custom attribute;
-					// see https://github.com/dotnet/runtime/blob/main/docs/design/features/covariant-return-methods.md.
-					if (preserveBaseOverridesAttribute != null)
-					{
-						return (xm != null && xm.IsDefined(preserveBaseOverridesAttribute, inherit: false) && y.IsAssignableFrom(x))
-						    || (ym != null && ym.IsDefined(preserveBaseOverridesAttribute, inherit: false) && x.IsAssignableFrom(y));
-					}
-
-					return false;
+					return IsCovariantReturnTypes(x, y, xm, ym);
 				}
 			}
 			return true;
+
+			static bool IsCovariantReturnTypes(Type x, Type y, MethodInfo xm = null, MethodInfo ym = null)
```

Comment:
Probably don't need default values for `xm` and `ym` since the function needs them to do something useful.